src/Storages/MergeTree/IntersectionsIndexes.h

```diff
@@ -0,0 +1,206 @@
+#pragma once
+
+#include <fmt/format.h>
+#include <Storages/MergeTree/RequestResponse.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+/// A boundary of a segment (left or right)
+struct PartToRead
+{
+    PartBlockRange range;
+    String name;
+
+    bool operator==(const PartToRead & rhs) const
+    {
+        return range == rhs.range && name == rhs.name;
+    }
+
+    bool operator<(const PartToRead & rhs) const
+    {
+        /// We allow only consecutive non-intersecting ranges
+        if (rhs.range.begin > range.begin && rhs.range.begin < range.end)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Bad ranges");
+        return range.begin < rhs.range.begin && range.end <= rhs.range.begin;
+    }
+};
+
+
+struct PartSegments
+{
+    enum class IntersectionResult
+    {
+        NO_INTERSECTION,
+        EXACTLY_ONE_INTERSECTION,
+        REJECT
+    };
+
+    void addPart(PartToRead part) { segments.insert(std::move(part)); }
+
+    IntersectionResult getIntersectionResult(PartToRead part)
+    {
+        bool intersected_before = false;
+        for (const auto & segment: segments)
+        {
+            auto are_intersect = [](auto & x, auto & y)
+            {
+                /// <= is important here, because we are working with segments [a, b]
+                if ((x.begin <= y.begin) && (y.begin <= x.end))
+                    return true;
+                if ((y.begin <= x.begin) && (x.begin <= y.end))
+                    return true;
+                return false;
+            };
```

Comment:
Why don't we check the case `[ { ] }`, but only `[ { } ]`?

Reply:
Because it is enough to check if two segments intersect. Literally two segments are intersect iff the beginning of one segments lies inside the other segment.